### ISSUE TITLE
SIGSEGV or SIGABRT during startup on Fedora CoreOS 33

### ISSUE THREAD

**Issue by zaphod42** [DataDog/datadog-agent#7033]
**Output of the info page (if this is a bug)**

As the agent fails to start the status page doesn't seem to be available.
```
2020-12-22 10:00:29 UTC | CORE | INFO | (pkg/util/log/log.go:465 in func1) | Found 'DD_PROXY_HTTPS' env var, using it for the Agent proxy settings
Could not reach agent: Get "https://localhost:5001/agent/status": dial tcp 127.0.0.1:5001: connect: connection refused
Make sure the agent is running before requesting the status and contact support if you continue having issues.
command terminated with exit code 255
Getting the status from the agent.
```

The agent is datadog/agent:7.23.1 with the following in the config file

```
logs:
  - type: journald
    path: /var/log/journal
    include_units:
      - kubelet.service
```

**Describe what happened:**

After rebuilding a kubernetes master node, which upgrade the underlying OS, the datadog-agent is no longer able to start on the node. The rebuild wipes the entire system, so there are no logs left behind from the prior install that could be interfering with the system.

I've seen the crash happen in two different ways. One is a SIGSEGV the other is a SIGABRT; however, both seem to be at the same place, which is a call to sd_journal_enumerate_data.

In the logs we see (after various other log tailing initialization):

```
2020-12-22 09:27:24 UTC | CORE | INFO | (pkg/logs/scheduler/scheduler.go:66 in Schedule) | Received a new logs config: journald
2020-12-22 09:27:24 UTC | CORE | INFO | (pkg/logs/input/journald/tailer.go:61 in Start) | Start tailing journal  /var/log/journal
Assertion 'output.pos >= size' failed at src/journal/compress.c:304, function decompress_blob_zstd(). Aborting.
SIGABRT: abort
PC=0x7f407b466c41 m=7 sigcode=18446744073709551610

goroutine 0 [idle]:
runtime: unknown pc 0x7f407b466c41
stack: frame={sp:0x7f405229e980, fp:0x0} stack=[0x7f4051a9f268,0x7f405229ee68)
00007f405229e880:  ffffffffffffffd4  00007f40380cefb8
...elided...
00007f405229ea70:  ffffffffffffffff  ffffffffffffffff
runtime: unknown pc 0x7f407b466c41
stack: frame={sp:0x7f405229e980, fp:0x0} stack=[0x7f4051a9f268,0x7f405229ee68)
00007f405229e880:  ffffffffffffffd4  00007f40380cefb8
00007f405229e890:  00007f40380d3978  00007f40042dda5b
...elided...
00007f405229ea70:  ffffffffffffffff  ffffffffffffffff

goroutine 580 [syscall]:
non-Go function
        pc=0x7f407b466c41
non-Go function
        pc=0x7f407b450536
non-Go function
        pc=0x7f40043aa87c
non-Go function
        pc=0x7f400438e3f6
non-Go function
        pc=0x7f4004395651
non-Go function
        pc=0x7f4004396b7d
runtime.cgocall(0x30bdd7a, 0xc000092d50, 0xc000698190)
        /root/.gimme/versions/go1.14.7.linux.amd64/src/runtime/cgocall.go:133 +0x5b fp=0xc000092d20 sp=0xc000092ce8 pc=0x13ba26b
github.com/coreos/go-systemd/sdjournal._Cfunc_my_sd_journal_enumerate_data(0x7f4004396a90, 0x7f4038014840, 0xc000698190, 0xc0004e10c0, 0x0)
        _cgo_gotypes.go:174 +0x4d fp=0xc000092d50 sp=0xc000092d20 pc=0x2f0a45d
github.com/coreos/go-systemd/sdjournal.(*Journal).GetEntry.func6(0x7f4004396a90, 0xc0008da030, 0xc000698190, 0xc0004e10c0, 0xc0014f93c8)
        /.omnibus/src/datadog-agent/src/github.com/DataDog/datadog-agent/vendor/github.com/coreos/go-systemd/sdjournal/journal.go:770 +0xda fp=0xc000092d90 sp=0xc000092d50 pc=0x2f
10d6a
github.com/coreos/go-systemd/sdjournal.(*Journal).GetEntry(0xc0008da030, 0x0, 0x0, 0x0)
        /.omnibus/src/datadog-agent/src/github.com/DataDog/datadog-agent/vendor/github.com/coreos/go-systemd/sdjournal/journal.go:770 +0x44a fp=0xc000092f18 sp=0xc000092d90 pc=0x2
f0d5da
github.com/DataDog/datadog-agent/pkg/logs/input/journald.(*Tailer).tail(0xc0006b0000)
        /.omnibus/src/datadog-agent/src/github.com/DataDog/datadog-agent/pkg/logs/input/journald/tailer.go:149 +0x121 fp=0xc000092fd8 sp=0xc000092f18 pc=0x2f13011
runtime.goexit()
        /root/.gimme/versions/go1.14.7.linux.amd64/src/runtime/asm_amd64.s:1373 +0x1 fp=0xc000092fe0 sp=0xc000092fd8 pc=0x141ec41
created by github.com/DataDog/datadog-agent/pkg/logs/input/journald.(*Tailer).Start
        /.omnibus/src/datadog-agent/src/github.com/DataDog/datadog-agent/pkg/logs/input/journald/tailer.go:62 +0x16e

```
or
```
2020-12-22 09:32:36 UTC | CORE | INFO | (pkg/logs/scheduler/scheduler.go:66 in Schedule) | Received a new logs config: journald
2020-12-22 09:32:36 UTC | CORE | INFO | (pkg/logs/input/journald/tailer.go:61 in Start) | Start tailing journal  /var/log/journal
fatal error: unexpected signal during runtime execution
[signal SIGSEGV: segmentation violation code=0x2 addr=0x7f1a4007b3d9 pc=0x7f1a7eaa26f5]

runtime stack:
runtime.throw(0x3922568, 0x2a)
        /root/.gimme/versions/go1.14.7.linux.amd64/src/runtime/panic.go:1116 +0x72
runtime.sigpanic()
        /root/.gimme/versions/go1.14.7.linux.amd64/src/runtime/signal_unix.go:679 +0x46a

goroutine 618 [syscall]:
non-Go function
        pc=0x7f1a7eaa26f5
non-Go function
        pc=0x7f1a0709e8ec
non-Go function
        pc=0x7f1a0714a11d
non-Go function
        pc=0x7f1a07151651
non-Go function
        pc=0x7f1a07152b7d
non-Go function
        pc=0x30bdb36
non-Go function
        pc=0x30bddbd
non-Go function
        pc=0x141e37f
runtime.cgocall(0x30bdd7a, 0xc00121ed50, 0xc0000102a0)
        /root/.gimme/versions/go1.14.7.linux.amd64/src/runtime/cgocall.go:133 +0x5b fp=0xc00121ed20 sp=0xc00121ece8 pc=0x13ba26b
github.com/coreos/go-systemd/sdjournal._Cfunc_my_sd_journal_enumerate_data(0x7f1a07152a90, 0x7f1a40003310, 0xc0000102a0, 0xc00073f1a0, 0x0)
        _cgo_gotypes.go:174 +0x4d fp=0xc00121ed50 sp=0xc00121ed20 pc=0x2f0a45d
github.com/coreos/go-systemd/sdjournal.(*Journal).GetEntry.func6(0x7f1a07152a90, 0xc000782000, 0xc0000102a0, 0xc00073f1a0, 0xc00104f198)
        /.omnibus/src/datadog-agent/src/github.com/DataDog/datadog-agent/vendor/github.com/coreos/go-systemd/sdjournal/journal.go:770 +0xda fp=0xc00121ed90 sp=0xc00121ed50 pc=0x2f10d6a
github.com/coreos/go-systemd/sdjournal.(*Journal).GetEntry(0xc000782000, 0x0, 0x0, 0x0)
        /.omnibus/src/datadog-agent/src/github.com/DataDog/datadog-agent/vendor/github.com/coreos/go-systemd/sdjournal/journal.go:770 +0x44a fp=0xc00121ef18 sp=0xc00121ed90 pc=0x2f0d5da
github.com/DataDog/datadog-agent/pkg/logs/input/journald.(*Tailer).tail(0xc000f95290)
        /.omnibus/src/datadog-agent/src/github.com/DataDog/datadog-agent/pkg/logs/input/journald/tailer.go:149 +0x121 fp=0xc00121efd8 sp=0xc00121ef18 pc=0x2f13011
runtime.goexit()
        /root/.gimme/versions/go1.14.7.linux.amd64/src/runtime/asm_amd64.s:1373 +0x1 fp=0xc00121efe0 sp=0xc00121efd8 pc=0x141ec41
created by github.com/DataDog/datadog-agent/pkg/logs/input/journald.(*Tailer).Start
        /.omnibus/src/datadog-agent/src/github.com/DataDog/datadog-agent/pkg/logs/input/journald/tailer.go:62 +0x16e
```

**Describe what you expected:**

No crash

**Steps to reproduce the issue:**

I assume it is a change in Fedora CoreOS 33, so try running on that OS.

**Additional environment details (Operating System, Cloud provider, etc):**

A few versions changed during the rebuild:
  * OS changed from Fedora CoreOS 32.20201104.3.0 to Fedora CoreOS 33.20201201.3.0
  * Kernel changed from 5.8.17-200.fc32.x86_64 to 5.9.11-200.fc33.x86_64
  * Docker changed from docker://19.3.11 to docker://19.3.13

Kubernetes version: v1.18.12

**Comment by pasanw** [comment]
We also faced the exact same issue with our production cluster, which runs Flatcar Linux.

Similar to @zaphod42, the issue surfaced only after an upgrade of the underlying CoreOS-based OS. In our case, it was Flatcar from 2605.6.0 to 2605.10.0.

For Flatcar, this entailed an upgrade of `systemd` from 245 to 246, in which "`systemd-journald` gained support for `zstd` compression of large fields in journal files". It's important to note that `systemd` is also upgraded from 245 to 246 from Fedora CoreOS 32 -> 33.

I was able to identify the cause of the issue. **It appears to be triggered when entries larger than the journald compression threshold (512 bytes by default) are added to the journal.** 

Simply creating a file larger than this (e.g. 600b) and pushing it to the journal (e.g. echo 600b | systemd-cat) can be used to reproduce this issue on a CoreOS-based host running systemd 246. Once this entry is made, the agent will continue to fail to start, exiting with one of the stack traces provided above. Clearing the journal archives (e.g. rm -rf /var/log/journal/*) would resolve the state.

A workaround is to disable compression entirely via journald.conf.

In summary, it appears that when journal entries are compressed with `zstd`, which happens starting with systemd 246, DD agent crashes when trying to read the journal. Hopefully DD can respond here and indicate whether the issue lies on their side or with the systemd/journald libraries.

**Comment by sgnn7** [comment]
Hi @zaphod and @pasanw - thank you for reporting this issue!

Just updating you that there is some work being done on this however I was unable to replicate this  (log entries showed correctly in Web UI) with:
- Ubuntu 21.04 (`systemd==247.1-4ubuntu1`) 
- With/without `SYSTEMD_JOURNAL_KEYED_HASH` flag 
- Using explicit `Compress=yes` in `/etc/systemd/journald.conf`
- Using `python3 -c 'print("b" * 7000)' | systemd-cat`

I'll try using FC33 next but it may be a bit more complex issue than just `systemd` `zstd` problem. It's possible Canonical has some extra patches but I will find out soon enough.

Since the underlying code used for reading these files is in a [third-party lib](github.com/coreos/go-systemd) and doesn't seem to have a matching bug/PR open for something similar, I am guessing that the fix may not be very simple (especially since it uses a [C lib](https://github.com/coreos/go-systemd/blob/408310d1ce3799ce411aa56d99b36250ae8b6aa7/sdjournal/journal.go#L770) for the parsing). If it is indeed as described in the comment from @pasanw, it may take some time to propagate to the end product and we may likely recommend the workaround suggested about disabling compression (we still need to verify the bug though).

We'll try to keep this issue updated as we work on this.

**Comment by sgnn7** [comment]
Small update - I didn't get a chance to keep digging much more into this this week (so no replication of the bug yet) but I'll continue next week. 

Some thinking on it though:
- When you run a containerized agent that reads the host's journald entries, you are using the container's C lib
- If the container's base image C lib reading is not compatible with host's journald writing style, this could easily cause incompatibilities.

@zaphod42 / @pasanw: Can you try running the agent 7.25.0 in the meantime? That docker base image comes with systemd `247.2` so it could in theory have the support for the `zstd` encoding.

**Comment by sgnn7** [comment]
More updates/notes from work on this yesterday:

The general data query path:
- Dockerized agent -> `coreos/gosystemd` Golang lib -> `libsystemd.so` -> `journald` reading on host

Fedora Core 33 reproduces the issue but only when running agent in Docker and pulling logs from host’s `journald`.
  - FC33 natively-installed agent works fine
  - FC33 uses systemd `libsystemd.so.0.29.0`, Agent docker image uses `libsystemd.so.0.30.0`

Other findings:
- Using a minimal `coreos/go-systemd/sdjournal` (the module we use to interface with the libsystemd) of the same version and with a small test file:
  - No apparent issues reading the `journald` from FC33 host (expected)
  - No apparent issues reading the host's `journald` from within the docker container (**not expected**)

I tried a bind-mounted `libsystemd.so.*` to maybe see if it could work with host's DLL but it wanted me to also link `glibc` which at that point is asking for too much of a risky approach to fixing this issue so I'm exploring other avenues still.

I'll keep the thread updated

**Comment by sgnn7** [comment]
Issue has been reproduced ([vagrant scripts](https://gist.github.com/sgnn7/e9586f7b8fe8f09eea021521ccb446d7)). Turning off compression in journald doesn't seem like it works as a workaround sadly.

**Comment by vickenty** [comment]
We believe the issue has been resolved in the latest release of Datadog agent. Please upgrade to version 7.28.1 to get the fix.

**Comment by vickenty** [comment]
Closing this as we believe the issue to be resolved. Please feel free to re-open if the problem surfaces again.
